In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Part I - Basic Data Exploration

Load the dataset

In [2]:
path_to_dataset = "googleplaystore.csv"
google_play_store_df = pd.read_csv(path_to_dataset)

Display first 10 samples, using `.head(`) method

In [3]:
google_play_store_df.head(n=10)

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   
5                         Paper flowers instructions  ART_AND_DESIGN     4.4   
6            Smoke Effect Photo Maker - Smoke Editor  ART_AND_DESIGN     3.8   
7                                   Infinite Painter  ART_AND_DESIGN     4.1   
8                               Garden Coloring Book  ART_AND_DESIGN     4.4   
9                      Kids Paint Free - Drawing Fun  ART_AND_DESIGN     4.7   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   
5     167  5.6M      50,000+  Free     0       Everyone   
6     178   19M      50,000+  Free     0       Everyone   
7   36815   29M   1,000,000+  Free     0       Everyone   
8   13791   33M   1,000,000+  Free     0       Everyone   
9     121  3.1M      10,000+  Free     0       Everyone   

                      Genres        Last Updated         Current Ver  \
0               Art & Design     January 7, 2018               1.0.0   
1  Art & Design;Pretend Play    January 15, 2018               2.0.0   
2               Art & Design      August 1, 2018               1.2.4   
3               Art & Design        June 8, 2018  Varies with device   
4    Art & Design;Creativity       June 20, 2018                 1.1   
5               Art & Design      March 26, 2017                 1.0   
6               Art & Design      April 26, 2018                 1.1   
7               Art & Design       June 14, 2018            6.1.61.1   
8               Art & Design  September 20, 2017               2.9.2   
9    Art & Design;Creativity        July 3, 2018                 2.8   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up  
5    2.3 and up  
6  4.0.3 and up  
7    4.2 and up  
8    3.0 and up  
9  4.0.3 and up

In EDA projects, I always like to see the column names using `.columns` attribute, rather than seeing it from table above.

In [4]:
google_play_store_df.columns.to_list()

['App',
 'Category',
 'Rating',
 'Reviews',
 'Size',
 'Installs',
 'Type',
 'Price',
 'Content Rating',
 'Genres',
 'Last Updated',
 'Current Ver',
 'Android Ver']

In [5]:
google_play_store_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


In [6]:
google_play_store_df.describe()

Rating
count  9367.000000
mean      4.193338
std       0.537431
min       1.000000
25%       4.000000
50%       4.300000
75%       4.500000
max      19.000000

In [7]:
num_samples = len(google_play_store_df)
print(f'There are {num_samples} samples in the dataset.')

There are 10841 samples in the dataset.


# Part II - Missing Values and Data Types

As I inspect the results of `google_play_store_df.head(n=10)` and `google_play_store_df.info()`, I observed the following:

- The first thing my eye caught is that there is only one feature kept as `float64` `(Rating`)

- Every other features are kept as `object`. I know that the pandas keep `str` datatype as `object` in `dataframes`. However, even though the `Price` is something that is numerical in our life, it stored as `object`. Moreover, there are more quantative features as well, like `Reviews`, `Size` and `Installs`. Thanks to the flexibility of Python, (if needed), I can convert those into numerical (e.g., I can define a function that maps the "19M" to 19.000.000.000).

- There are total of 10841 rows in the dataset. However, there are some missing values, based on the results of `.info()` method call. For example, there are 10841-9367=1474 missing values for `Rating` column.

## 2.1. Check missing values

In [8]:
# Display rows with any missing values
google_play_store_df[google_play_store_df.isna().any(axis=1)]

App             Category  Rating Reviews  \
15     Learn To Draw Kawaii Characters       ART_AND_DESIGN     3.2      55   
23              Mcqueen Coloring pages       ART_AND_DESIGN     NaN      61   
113          Wrinkles and rejuvenation               BEAUTY     NaN     182   
123             Manicure - nail design               BEAUTY     NaN     119   
126       Skin Care and Natural Beauty               BEAUTY     NaN     654   
...                                ...                  ...     ...     ...   
10824                        Cardio-FR              MEDICAL     NaN      67   
10825               Naruto & Boruto FR               SOCIAL     NaN       7   
10831         payermonstationnement.fr  MAPS_AND_NAVIGATION     NaN      38   
10835                         FR Forms             BUSINESS     NaN       0   
10838           Parkinson Exercices FR              MEDICAL     NaN       3   

       Size  Installs  Type Price Content Rating  \
15     2.7M    5,000+  Free     0       Everyone   
23     7.0M  100,000+  Free     0       Everyone   
113    5.7M  100,000+  Free     0   Everyone 10+   
123    3.7M   50,000+  Free     0       Everyone   
126    7.4M  100,000+  Free     0           Teen   
...     ...       ...   ...   ...            ...   
10824   82M   10,000+  Free     0       Everyone   
10825  7.7M      100+  Free     0           Teen   
10831  9.8M    5,000+  Free     0       Everyone   
10835  9.6M       10+  Free     0       Everyone   
10838  9.5M    1,000+  Free     0       Everyone   

                                Genres        Last Updated Current Ver  \
15                        Art & Design        June 6, 2018         NaN   
23     Art & Design;Action & Adventure       March 7, 2018       1.0.0   
113                             Beauty  September 20, 2017         8.0   
123                             Beauty       July 23, 2018         1.3   
126                             Beauty       July 17, 2018        1.15   
...                                ...                 ...         ...   
10824                          Medical       July 31, 2018       2.2.2   
10825                           Social    February 2, 2018         1.0   
10831                Maps & Navigation       June 13, 2018   2.0.148.0   
10835                         Business  September 29, 2016       1.1.5   
10838                          Medical    January 20, 2017         1.0   

      Android Ver  
15     4.2 and up  
23     4.1 and up  
113    3.0 and up  
123    4.1 and up  
126    4.1 and up  
...           ...  
10824  4.4 and up  
10825  4.0 and up  
10831  4.0 and up  
10835  4.0 and up  
10838  2.2 and up  

[1481 rows x 13 columns]

## 2.2. Identify data types

In [9]:
google_play_store_df.dtypes

App                object
Category           object
Rating            float64
Reviews            object
Size               object
Installs           object
Type               object
Price              object
Content Rating     object
Genres             object
Last Updated       object
Current Ver        object
Android Ver        object
dtype: object

2.3. Check if each column has an appropriate data type.

In [10]:
# In order to display dtypes, I again invoke .info() method on my df
google_play_store_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


• The first thing my eye caught is that there is only one feature kept as `float64` (Rating)


• Every other features are kept as `object`. I know that the pandas keep `str` datatype as `object` in dataframes. However, even though the Price is something that is numerical in our life, it stored as `object`. Moreover, there are more quantative features as well, like Reviews, Size and Installs. Thanks to the flexibility of Python, (if needed), I can convert those into numerical (e.g., I can define a function that maps the "19M" to 19.000.000.000).

In [11]:
# Display unique values for each column in the dataset
for col in google_play_store_df.columns:
  print(f'Unique values for feature {col}:\n{google_play_store_df[col].unique()}\n----------------------------------------------------------------------\n')

Unique values for feature App:
['Photo Editor & Candy Camera & Grid & ScrapBook' 'Coloring book moana'
 'U Launcher Lite – FREE Live Cool Themes, Hide Apps' ...
 'Parkinson Exercices FR' 'The SCP Foundation DB fr nn5n'
 'iHoroscope - 2018 Daily Horoscope & Astrology']
----------------------------------------------------------------------

Unique values for feature Category:
['ART_AND_DESIGN' 'AUTO_AND_VEHICLES' 'BEAUTY' 'BOOKS_AND_REFERENCE'
 'BUSINESS' 'COMICS' 'COMMUNICATION' 'DATING' 'EDUCATION' 'ENTERTAINMENT'
 'EVENTS' 'FINANCE' 'FOOD_AND_DRINK' 'HEALTH_AND_FITNESS' 'HOUSE_AND_HOME'
 'LIBRARIES_AND_DEMO' 'LIFESTYLE' 'GAME' 'FAMILY' 'MEDICAL' 'SOCIAL'
 'SHOPPING' 'PHOTOGRAPHY' 'SPORTS' 'TRAVEL_AND_LOCAL' 'TOOLS'
 'PERSONALIZATION' 'PRODUCTIVITY' 'PARENTING' 'WEATHER' 'VIDEO_PLAYERS'
 'NEWS_AND_MAGAZINES' 'MAPS_AND_NAVIGATION' '1.9']
----------------------------------------------------------------------

Unique values for feature Rating:
[ 4.1  3.9  4.7  4.5  4.3  4.4  3.8  4.2  4.6

I noticed that for Price column, there is 'Everyone' when I looked at the unique values in above.

In [12]:
google_play_store_df[google_play_store_df["Price"] == "Everyone"]

App Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

However, Price is not the only problematic column for 10472nd sample. Also category seems wrong, Type is 0, Size is 1,000+, etc. Actually, the only missing thing in here is Category. Because all values except for App column seem kind of like shifted left. But still, I prefer to remove this sample.

In [13]:
# Find the invalid of wrong sample
index_of_invalid_sample = google_play_store_df.index[google_play_store_df["Price"] == "Everyone"].to_list()[0]

# Remove this invalid sample from dataframe
google_play_store_df = google_play_store_df.drop(index_of_invalid_sample, axis='index')

In [14]:
# Change dtype of Reviews column to int
google_play_store_df["Reviews"] = google_play_store_df["Reviews"].astype(int)

In [16]:
google_play_store_df.head(3)

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   

   Reviews  Size    Installs  Type Price Content Rating  \
0      159   19M     10,000+  Free     0       Everyone   
1      967   14M    500,000+  Free     0       Everyone   
2    87510  8.7M  5,000,000+  Free     0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up  
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up  
2               Art & Design    August 1, 2018       1.2.4  4.0.3 and up

In [17]:
google_play_store_df["Size"].unique()

array(['19M', '14M', '8.7M', '25M', '2.8M', '5.6M', '29M', '33M', '3.1M',
       '28M', '12M', '20M', '21M', '37M', '2.7M', '5.5M', '17M', '39M',
       '31M', '4.2M', '7.0M', '23M', '6.0M', '6.1M', '4.6M', '9.2M',
       '5.2M', '11M', '24M', 'Varies with device', '9.4M', '15M', '10M',
       '1.2M', '26M', '8.0M', '7.9M', '56M', '57M', '35M', '54M', '201k',
       '3.6M', '5.7M', '8.6M', '2.4M', '27M', '2.5M', '16M', '3.4M',
       '8.9M', '3.9M', '2.9M', '38M', '32M', '5.4M', '18M', '1.1M',
       '2.2M', '4.5M', '9.8M', '52M', '9.0M', '6.7M', '30M', '2.6M',
       '7.1M', '3.7M', '22M', '7.4M', '6.4M', '3.2M', '8.2M', '9.9M',
       '4.9M', '9.5M', '5.0M', '5.9M', '13M', '73M', '6.8M', '3.5M',
       '4.0M', '2.3M', '7.2M', '2.1M', '42M', '7.3M', '9.1M', '55M',
       '23k', '6.5M', '1.5M', '7.5M', '51M', '41M', '48M', '8.5M', '46M',
       '8.3M', '4.3M', '4.7M', '3.3M', '40M', '7.8M', '8.8M', '6.6M',
       '5.1M', '61M', '66M', '79k', '8.4M', '118k', '44M', '695k', '1.6M',
     